[WIP] use Go's native t.TempDir() in tests

thaJeztah opened the pull request:
follow-up to https://github.com/moby/moby/pull/43340

thaJeztah committed:
use Go native t.TempDir() in tests

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Couple of fixes to make, looks like;

```go
=== RUN   TestSysctlOverride
    testing.go:967: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverride4168880149/001/mounts/shm: device or resource busy

=== CONT  TestSetGetMeta
db_test.go:15: assertion failed: error is not nil: open /tmp/TestSetGetMeta1284580809/001: is a directory
--- FAIL: TestSetGetMeta (0.00s)
````

Also windows being windows, I guess:

```
=== RUN   TestLoadOrCreateTrustKeyInvalidKeyFile
testing.go:967: TempDir RemoveAll cleanup: remove C:\Users\ContainerAdministrator\AppData\Local\Temp\TestLoadOrCreateTrustKeyInvalidKeyFile1320838553\001\keyfile3365672289: The process cannot access the file because it is being used by another process.

=== RUN   TestIsEmptyDir
testing.go:967: TempDir RemoveAll cleanup: remove C:\Users\ContainerAdministrator\AppData\Local\Temp\TestIsEmptyDir1283921867\001\dir-with-empty-file\file369182996: The process cannot access the file because it is being used by another process.

=== CONT  TestRestore
testing.go:967: TempDir RemoveAll cleanup: remove C:\Users\ContainerAdministrator\AppData\Local\Temp\TestRestore147916941\001\volumes\metadata.db: The process cannot access the file because it is being used by another process.
```

tianon commented:
FWIW, this one helps me feel better about #43340 :sweat_smile:

Somehow using `t.TempDir()` feels less "magical" than having something else "automatically" adding things to Go's cleanup callbacks. :shrug: :innocent:

thaJeztah commented:
Ah! I see you found this one as well; yes, I like `t.TempDir()`, and think we should phase-out the `fs` things for _most_ uses (there are a couple of cases where it could still be useful, but we should have a close look if we need a utility for those, or if we could just as well use some (minimal) amount of "regular" go code in the test.

thaJeztah commented:
Couple of failures; need to look what's special about those tests and/or if there's possibly a bug in the tests;


On Windows:

```
=== Failed
=== FAIL: github.com/docker/docker/daemon TestLoadOrCreateTrustKeyInvalidKeyFile (0.00s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestLoadOrCreateTrustKeyInvalidKeyFile2014634395\001\keyfile4156691647: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/daemon/graphdriver TestIsEmptyDir (0.01s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestIsEmptyDir1962964337\001\dir-with-empty-file\file2523853824: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/pkg/directory TestSizeEmptyFile (0.00s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestSizeEmptyFile1562416712\001\file16507846: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/pkg/directory TestSizeNonemptyFile (0.00s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestSizeNonemptyFile1240832785\001\file3265662846: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/pkg/directory TestSizeFileAndNestedDirectoryEmpty (0.00s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestSizeFileAndNestedDirectoryEmpty2163416550\001\file3715413181: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/pkg/directory TestSizeFileAndNestedDirectoryNonempty (0.00s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestSizeFileAndNestedDirectoryNonempty878205470\001\file3280422273: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/volume/service TestSetGetMeta (0.01s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestSetGetMeta3332268057\001\db: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/volume/service TestList (0.03s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestList2846947953\001\volumes\metadata.db: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/volume/service TestRestore (0.02s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestRestore3368254142\001\volumes\metadata.db: The process cannot access the file because it is being used by another process.
```

On rootless:

```
=== RUN   TestRemoveContainerWithRemovedVolume
    remove_test.go:37: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume868372652/001': mkdir /tmp/TestRemoveContainerWithRemovedVolume868372652/001: permission denied
--- FAIL: TestRemoveContainerWithRemovedVolume (0.10s)
```

On arm64:

```
=== RUN   TestSysctlOverride
    testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverride2860094781/001/mounts/shm: device or resource busy
--- FAIL: TestSysctlOverride (0.00s)
```

thaJeztah committed:
Fix TestSysctlOverride tests

    === RUN   TestSysctlOverride
        testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverride2860094781/001/mounts/shm: device or resource busy
    --- FAIL: TestSysctlOverride (0.00s)

fix panic in TestTmpfsDevShmNoDupMount

=== RUN   TestTmpfsDevShmNoDupMount
    --- FAIL: TestTmpfsDevShmNoDupMount (0.01s)
    panic: runtime error: invalid memory address or nil pointer dereference [recovered]
        panic: runtime error: invalid memory address or nil pointer dereference
    [signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1209078]
    goroutine 110 [running]:
    testing.tRunner.func1.2({0x180e0e0, 0x2af5180})
        /usr/local/go/src/testing/testing.go:1389 +0x1c8
    testing.tRunner.func1()
        /usr/local/go/src/testing/testing.go:1392 +0x380
    panic({0x180e0e0, 0x2af5180})
        /usr/local/go/src/runtime/panic.go:838 +0x20c
    github.com/docker/docker/plugin.(*Store).getAllByCap(0x0?, {0x1aa19ca?, 0x4000963d48?})
        /go/src/github.com/docker/docker/plugin/store.go:76 +0x38
    github.com/docker/docker/plugin.(*Store).GetAllManagedPluginsByCap(...)
        /go/src/github.com/docker/docker/plugin/store.go:178
    github.com/docker/docker/daemon.(*Daemon).cleanupMetricsPlugins(0x4000169680)
        /go/src/github.com/docker/docker/daemon/metrics.go:113 +0x7c
    github.com/docker/docker/daemon.(*Daemon).Shutdown(0x4000169680)
        /go/src/github.com/docker/docker/daemon/daemon.go:1226 +0x5f8
    github.com/docker/docker/daemon.setupFakeDaemon.func1()
        /go/src/github.com/docker/docker/daemon/oci_linux_test.go:50 +0x34
    testing.(*common).Cleanup.func1()
        /usr/local/go/src/testing/testing.go:1034 +0x10c
    testing.(*common).runCleanup(0x4000ac44e0, 0x1dba270?)
        /usr/local/go/src/testing/testing.go:1203 +0x9c
    testing.tRunner.func2()
        /usr/local/go/src/testing/testing.go:1433 +0x30
    testing.tRunner(0x4000ac44e0, 0x1b5a968)
        /usr/local/go/src/testing/testing.go:1445 +0x13c
    created by testing.(*T).Run
        /usr/local/go/src/testing/testing.go:1486 +0x328

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

daemon: fix panic in test

=== Failed
=== FAIL: daemon TestTmpfsDevShmNoDupMount (0.00s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x199dc18]

goroutine 566 [running]:
testing.tRunner.func1.2({0x1bd9000, 0x33c1970})
	/usr/local/go/src/testing/testing.go:1632 +0x1bc
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1635 +0x334
panic({0x1bd9000?, 0x33c1970?})
	/usr/local/go/src/runtime/panic.go:791 +0x124
github.com/docker/docker/daemon.(*fakeImageService).Cleanup(0x0?)
	<autogenerated>:1 +0x28
github.com/docker/docker/daemon.(*Daemon).Shutdown(0x40006b26c8, {0x229def8, 0x3540a40})
	/go/src/github.com/docker/docker/daemon/daemon.go:1307 +0x644
github.com/docker/docker/daemon.setupFakeDaemon.func1()
	/go/src/github.com/docker/docker/daemon/oci_linux_test.go:62 +0x40
testing.(*common).Cleanup.func1()
	/usr/local/go/src/testing/testing.go:1176 +0xf8
testing.(*common).runCleanup(0x400057d040, 0x2270000?)
	/usr/local/go/src/testing/testing.go:1354 +0xd0
testing.tRunner.func2()
	/usr/local/go/src/testing/testing.go:1684 +0x2c
testing.tRunner(0x400057d040, 0x20429f8)
	/usr/local/go/src/testing/testing.go:1696 +0x110
created by testing.(*T).Run in goroutine 1
	/usr/local/go/src/testing/testing.go:1743 +0x314

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Looking better; 3 remaining ones on Windows;

```
=== FAIL: github.com/docker/docker/daemon/graphdriver TestIsEmptyDir (0.00s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestIsEmptyDir2823795693\001\dir-with-empty-file\file2625561089: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/pkg/directory TestSizeFileAndNestedDirectoryNonempty (0.00s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestSizeFileAndNestedDirectoryNonempty4246252950\001\nested3442260313\file21164327: The process cannot access the file because it is being used by another process.

=== FAIL: github.com/docker/docker/volume/service TestRestore (0.03s)
    testing.go:1090: TempDir RemoveAll cleanup: remove C:\Users\CONTAI~1\AppData\Local\Temp\TestRestore618321306\001\volumes\metadata.db: The process cannot access the file because it is being used by another process.
```

More failures on Linux though;

rootless:

```
=== RUN   TestRemoveContainerWithRemovedVolume
    remove_test.go:39: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume1312590373/001': mkdir /tmp/TestRemoveContainerWithRemovedVolume1312590373/001: permission denied
--- FAIL: TestRemoveContainerWithRemovedVolume (0.13s)
```

```
=== RUN   TestTmpfsDevShmNoDupMount
--- FAIL: TestTmpfsDevShmNoDupMount (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x1209078]
goroutine 110 [running]:
testing.tRunner.func1.2({0x180e0e0, 0x2af5180})
	/usr/local/go/src/testing/testing.go:1389 +0x1c8
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1392 +0x380
panic({0x180e0e0, 0x2af5180})
	/usr/local/go/src/runtime/panic.go:838 +0x20c
github.com/docker/docker/plugin.(*Store).getAllByCap(0x0?, {0x1aa19ca?, 0x4000963d48?})
	/go/src/github.com/docker/docker/plugin/store.go:76 +0x38
github.com/docker/docker/plugin.(*Store).GetAllManagedPluginsByCap(...)
	/go/src/github.com/docker/docker/plugin/store.go:178
github.com/docker/docker/daemon.(*Daemon).cleanupMetricsPlugins(0x4000169680)
	/go/src/github.com/docker/docker/daemon/metrics.go:113 +0x7c
github.com/docker/docker/daemon.(*Daemon).Shutdown(0x4000169680)
	/go/src/github.com/docker/docker/daemon/daemon.go:1226 +0x5f8
github.com/docker/docker/daemon.setupFakeDaemon.func1()
	/go/src/github.com/docker/docker/daemon/oci_linux_test.go:50 +0x34
testing.(*common).Cleanup.func1()
	/usr/local/go/src/testing/testing.go:1034 +0x10c
testing.(*common).runCleanup(0x4000ac44e0, 0x1dba270?)
	/usr/local/go/src/testing/testing.go:1203 +0x9c
testing.tRunner.func2()
	/usr/local/go/src/testing/testing.go:1433 +0x30
testing.tRunner(0x4000ac44e0, 0x1b5a968)
	/usr/local/go/src/testing/testing.go:1445 +0x13c
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1486 +0x328
```

Hmm... and no details on these (perhaps they're because of the one above)

```
Failed
Stacktrace
=== RUN   TestVerifyPlatformContainerResources
=== PAUSE TestVerifyPlatformContainerResources

Failed
Stacktrace
=== RUN   TestGetBlkioWeightDevices
=== PAUSE TestGetBlkioWeightDevices

Failed
Stacktrace
=== RUN   TestRootMountCleanup
=== PAUSE TestRootMountCleanup

Failed
Stacktrace
=== RUN   TestGetBlkioThrottleDevices
=== PAUSE TestGetBlkioThrottleDevices
```

thaJeztah commented:
Getting closer; Windows is green now!

rootless is still failing on this one;

```
=== RUN   TestRemoveContainerWithRemovedVolume
    remove_test.go:39: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume1349239570/001': mkdir /tmp/TestRemoveContainerWithRemovedVolume1349239570/001: permission denied
--- FAIL: TestRemoveContainerWithRemovedVolume (0.11s)
```

Linux still has some failures;

```
=== RUN   TestSysctlOverride
    testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverride3702360633/001/mounts/shm: device or resource busy
--- FAIL: TestSysctlOverride (0.00s)

=== RUN   TestSysctlOverrideHost
    testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverrideHost226485533/001/mounts/shm: device or resource busy
--- FAIL: TestSysctlOverrideHost (0.00s)


=== RUN   TestDockerSuite/TestCpUnprivilegedUser
    docker_cli_cp_test.go:339: assertion failed: 
        Command:  su unprivilegeduser -c /usr/local/cli/docker cp 28f1d182a8d77c07c85dadf9ecbb4d7d0c9369042acfd66cec671a13423ebb60:test /tmp/TestDockerSuiteTestCpUnprivilegedUser2002826160/001
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   lstat /tmp/TestDockerSuiteTestCpUnprivilegedUser2002826160/001: permission denied
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerSuite/TestCpUnprivilegedUser (0.71s)

=== RUN   TestDockerSuite/TestRunWithVolumesIsRecursive
    testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestDockerSuiteTestRunWithVolumesIsRecursive1156692230/001/tmpfs: device or resource busy
    --- FAIL: TestDockerSuite/TestRunWithVolumesIsRecursive (0.49s)


=== RUN   TestDockerSuite/TestSaveRepoWithMultipleImages
    docker_cli_save_load_test.go:228: assertion failed: error is not nil: exit status 1
    --- FAIL: TestDockerSuite/TestSaveRepoWithMultipleImages (1.12s)
```

thaJeztah committed:
Try fixing unmounts in TestSysctlxxx

    [2022-06-01T12:07:16.248Z] === FAIL: daemon TestSysctlOverride (0.00s)
    [2022-06-01T12:07:16.248Z]     testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverride982023519/001/mounts/shm: device or resource busy
    [2022-06-01T12:07:16.248Z]
    [2022-06-01T12:07:16.248Z] === FAIL: daemon TestSysctlOverrideHost (0.00s)
    [2022-06-01T12:07:16.248Z]     testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverrideHost881080732/001/mounts/shm: device or resource busy

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Remaining ones on Windows (still debugging this one);

```
=== FAIL: github.com/docker/docker/pkg/archive TestUntarPathWithInvalidDest (0.04s)
    archive_test.go:300: srcFile: C:\Users\CONTAI~1\AppData\Local\Temp\TestUntarPathWithInvalidDest212983393\001\src
    archive_test.go:301: tarFile: C:\Users\CONTAI~1\AppData\Local\Temp\TestUntarPathWithInvalidDest212983393\001\src.tar
    archive_test.go:306: srcFileU: /tmp/src
    archive_test.go:307: tarFileU: /tmp/src.tar
    archive_test.go:311: assertion failed: error is not nil: exit status 2: tar: Removing leading `/' from member names
        tar: /tmp/src: Cannot stat: No such file or directory
        tar: Exiting with failure status due to previous errors
        

=== FAIL: github.com/docker/docker/pkg/archive TestUntarPath (0.04s)
    archive_test.go:352: assertion failed: error is not nil: exit status 2: tar: /tmp/001/src.tar: Cannot open: No such file or directory
        tar: Error is not recoverable: exiting now
        

=== FAIL: github.com/docker/docker/pkg/archive TestUntarPathWithDestinationFile (0.04s)
    archive_test.go:384: assertion failed: error is not nil: exit status 2: tar: /tmp/001/src.tar: Cannot open: No such file or directory
        tar: Error is not recoverable: exiting now
        

=== FAIL: github.com/docker/docker/pkg/archive TestUntarPathWithDestinationSrcFileAsFolder (0.04s)
    archive_test.go:419: assertion failed: error is not nil: exit status 2: tar: /tmp/001/src.tar: Cannot open: No such file or directory
        tar: Error is not recoverable: exiting now
```

rootless:

```
=== RUN   TestRemoveContainerWithRemovedVolume
    remove_test.go:41: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume616613276/001': mkdir /tmp/TestRemoveContainerWithRemovedVolume616613276/001: permission denied
--- FAIL: TestRemoveContainerWithRemovedVolume (0.12s)
```

Linux:

```
=== RUN   TestDockerSuite/TestCpUnprivilegedUser
    docker_cli_cp_test.go:339: assertion failed: 
        Command:  su unprivilegeduser -c /usr/local/cli/docker cp b51b3e12fb57157451ae5515834a1341c73baefe74bcc5b89bb0ab50aff75a4f:test /tmp/TestDockerSuiteTestCpUnprivilegedUser4016000516/001
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   lstat /tmp/TestDockerSuiteTestCpUnprivilegedUser4016000516/001: permission denied
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerSuite/TestCpUnprivilegedUser (0.64s)

=== RUN   TestDockerSuite/TestSaveRepoWithMultipleImages
    docker_cli_save_load_test.go:228: assertion failed: error is not nil: exit status 1
    --- FAIL: TestDockerSuite/TestSaveRepoWithMultipleImages (1.79s)
```

arm64 (different version of Linux?)

```
=== RUN   TestSysctlOverride
    oci_linux_test.go:61: umount /tmp/TestSysctlOverride2641690171/001/containers/fakecontainer/mount/shm, flags: 0x2: no such file or directory
    oci_linux_test.go:66: unlinkat /tmp/TestSysctlOverride2641690171/001/containers/fakecontainer/mounts/shm: device or resource busy
    testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverride2641690171/001/containers/fakecontainer/mounts/shm: device or resource busy
--- FAIL: TestSysctlOverride (0.00s)

=== RUN   TestSysctlOverrideHost
    oci_linux_test.go:61: umount /tmp/TestSysctlOverrideHost1378173456/001/containers/fakecontainer/mount/shm, flags: 0x2: no such file or directory
    oci_linux_test.go:66: unlinkat /tmp/TestSysctlOverrideHost1378173456/001/containers/fakecontainer/mounts/shm: device or resource busy
    testing.go:1090: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverrideHost1378173456/001/containers/fakecontainer/mounts/shm: device or resource busy
--- FAIL: TestSysctlOverrideHost (0.00s)

```

thaJeztah commented:
Failing on integration; rootless;

```
=== FAIL: amd64.integration.container TestRemoveContainerWithRemovedVolume (0.08s)
    remove_test.go:42: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume160279588/001': mkdir /tmp/TestRemoveContainerWithRemovedVolume160279588/001: permission denied
```

Failing on integration; systemd;

```
=== Failed
=== FAIL: amd64.integration.plugin.authz TestAuthZPluginV2RejectVolumeRequests (5.74s)
    authz_plugin_v2_test.go:99: assertion failed: error is not nil: Error response from daemon: plugin "riyaz/authz-no-volume-plugin:latest" not found
```


Failing unit-test:

```
=== FAIL: daemon TestSysctlOverrideHost (0.00s)
    oci_linux_test.go:60: umount /tmp/TestSysctlOverrideHost1847643613/001/containers/fakecontainer/mount/shm, flags: 0x2: no such file or directory
    oci_linux_test.go:65: unlinkat /tmp/TestSysctlOverrideHost1847643613/001/containers/fakecontainer/mounts/shm: device or resource busy
    testing.go:1097: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverrideHost1847643613/001/containers/fakecontainer/mounts/shm: device or resource busy
```


I may split this one up in smaller chunks, and leave the ones that need more attention.

thaJeztah commented:
arm64 (linux) failures;

```
=== FAIL: daemon TestSysctlOverride (0.00s)
    oci_linux_test.go:61: umount /tmp/TestSysctlOverride3979695660/001/containers/fakecontainer/mount/shm, flags: 0x2: no such file or directory
    oci_linux_test.go:66: unlinkat /tmp/TestSysctlOverride3979695660/001/containers/fakecontainer/mounts/shm: device or resource busy
    testing.go:1097: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverride3979695660/001/containers/fakecontainer/mounts/shm: device or resource busy

=== FAIL: daemon TestSysctlOverrideHost (0.00s)
    oci_linux_test.go:61: umount /tmp/TestSysctlOverrideHost1676261752/001/containers/fakecontainer/mount/shm, flags: 0x2: no such file or directory
    oci_linux_test.go:66: unlinkat /tmp/TestSysctlOverrideHost1676261752/001/containers/fakecontainer/mounts/shm: device or resource busy
    testing.go:1097: TempDir RemoveAll cleanup: unlinkat /tmp/TestSysctlOverrideHost1676261752/001/containers/fakecontainer/mounts/shm: device or resource busy

=== FAIL: pkg/idtools TestNewIDMappings (0.20s)
    idtools_unix_test.go:418: assertion failed: error is not nil: exit status 2: Unable to access /tmp/TestNewIDMappings630910663/001 directory with user UID:165536 and GID:165536:
        ls: cannot access '/tmp/TestNewIDMappings630910663/001': Permission denied
```


These are failing

```
=== FAIL: amd64.integration-cli TestDockerPluginSuite/TestPluginEnableDisableNegative (15.04s)
    docker_cli_plugins_test.go:242: assertion failed: error is not nil: 
        Command:  /usr/local/cli/docker plugin install --grant-all-permissions 127.0.0.1:5000/plugin/basic:latest
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   Error response from daemon: error resolving plugin reference: failed to do request: Head "https://127.0.0.1:5000/v2/plugin/basic/manifests/latest": http: server gave HTTP response to HTTPS client
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerPluginSuite/TestPluginEnableDisableNegative (15.04s)

=== FAIL: amd64.integration-cli TestDockerPluginSuite/TestPluginForceRemove (15.04s)
    docker_cli_plugins_test.go:77: assertion failed: error is not nil: 
        Command:  /usr/local/cli/docker plugin install --grant-all-permissions 127.0.0.1:5000/plugin/basic:latest
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   Error response from daemon: error resolving plugin reference: failed to do request: Head "https://127.0.0.1:5000/v2/plugin/basic/manifests/latest": http: server gave HTTP response to HTTPS client

        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerPluginSuite/TestPluginForceRemove (15.04s)

=== FAIL: amd64.integration-cli TestDockerPluginSuite/TestPluginInspect (15.04s)
    docker_cli_plugins_test.go:287: assertion failed: error is not nil: 
        Command:  /usr/local/cli/docker plugin install --grant-all-permissions 127.0.0.1:5000/plugin/basic:latest
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   Error response from daemon: error resolving plugin reference: failed to do request: Head "https://127.0.0.1:5000/v2/plugin/basic/manifests/latest": http: server gave HTTP response to HTTPS client

        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerPluginSuite/TestPluginInspect (15.04s)

=== FAIL: amd64.integration-cli TestDockerPluginSuite/TestPluginInstallDisable (15.03s)
    docker_cli_plugins_test.go:137: assertion failed: error is not nil: 
        Command:  /usr/local/cli/docker plugin install --grant-all-permissions --disable 127.0.0.1:5000/plugin/basic:latest
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   Error response from daemon: error resolving plugin reference: failed to do request: Head "https://127.0.0.1:5000/v2/plugin/basic/manifests/latest": http: server gave HTTP response to HTTPS client
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerPluginSuite/TestPluginInstallDisable (15.03s)

=== FAIL: amd64.integration-cli TestDockerPluginSuite (173.77s)
```

```
=== FAIL: amd64.integration-cli TestDockerCLICpSuite/TestCpUnprivilegedUser (0.43s)
    docker_cli_cp_test.go:345: assertion failed: 
        Command:  su unprivilegeduser -c /usr/local/cli/docker cp b7fb550358fd26b00db00ec4b3bd6be4d2af71ec6c866c1461d2ac600694a7fb:test /tmp/TestDockerCLICpSuiteTestCpUnprivilegedUser1344357240/001
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   lstat /tmp/TestDockerCLICpSuiteTestCpUnprivilegedUser1344357240/001: permission denied

        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerCLICpSuite/TestCpUnprivilegedUser (0.43s)
=== FAIL: amd64.integration-cli TestDockerCLICpSuite (25.70s)
```

Rootless;

```
=== FAIL: amd64.integration.container TestRemoveContainerWithRemovedVolume (0.08s)
    remove_test.go:41: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume227027773/001/hostPath': mkdir /tmp/TestRemoveContainerWithRemovedVolume227027773/001: permission denied
```

Interesting why it tries to create `/tmp/TestRemoveContainerWithRemovedVolume227027773/001`, and not `/tmp/TestRemoveContainerWithRemovedVolume227027773/001/hostPath`

Windows failing on the same test, but in a different spot;

```
=== FAIL: github.com/docker/docker/integration/container TestRemoveContainerWithRemovedVolume (0.26s)
    container.go:52: assertion failed: error is not nil: Error response from daemon: invalid volume specification: 'C:\Users\RUNNER~1\AppData\Local\Temp\TestRemoveContainerWithRemovedVolume724343939\001\hostPath:c:\test': invalid mount config for type "bind": bind source path does not exist: c:\users\runner~1\appdata\local\temp\testremovecontainerwithremovedvolume724343939\001\hostpath
```

Some failures in pkg/archive (but may be related to the debugging commit I added) on Windows 2022;

```
=== FAIL: github.com/docker/docker/pkg/archive TestUntarPathWithInvalidDest (0.05s)
    archive_test.go:301: srcFile: C:\Users\CONTAI~1\AppData\Local\Temp\TestUntarPathWithInvalidDest698436617\001\src
    archive_test.go:302: tarFile: C:\Users\CONTAI~1\AppData\Local\Temp\TestUntarPathWithInvalidDest698436617\001\src.tar
    archive_test.go:307: srcFileU: /tmp/src
    archive_test.go:308: tarFileU: /tmp/src.tar
    archive_test.go:312: assertion failed: error is not nil: exit status 2: tar: Removing leading `/' from member names
        tar: /tmp/src: Cannot stat: No such file or directory
        tar: Exiting with failure status due to previous errors
        

=== FAIL: github.com/docker/docker/pkg/archive TestUntarPath (0.05s)
    archive_test.go:353: assertion failed: error is not nil: exit status 2: tar: /tmp/001/src.tar: Cannot open: No such file or directory
        tar: Error is not recoverable: exiting now
        

=== FAIL: github.com/docker/docker/pkg/archive TestUntarPathWithDestinationFile (0.05s)
    archive_test.go:385: assertion failed: error is not nil: exit status 2: tar: /tmp/001/src.tar: Cannot open: No such file or directory
        tar: Error is not recoverable: exiting now
        

=== FAIL: github.com/docker/docker/pkg/archive TestUntarPathWithDestinationSrcFileAsFolder (0.05s)
    archive_test.go:420: assertion failed: error is not nil: exit status 2: tar: /tmp/001/src.tar: Cannot open: No such file or directory
        tar: Error is not recoverable: exiting now
```

thaJeztah committed:
fix TestRemoveContainerWithRemovedVolume

Failing on rootless; probably because the parent dir doesn't have
the right permissions;

    === FAIL: amd64.integration.container TestRemoveContainerWithRemovedVolume (0.07s)
        remove_test.go:41: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume1924909400/001/hostPath': mkdir /tmp/TestRemoveContainerWithRemovedVolume1924909400/001: permission denied

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah committed:
fix: TestDockerCLICpSuite/TestCpUnprivilegedUser

See if this works to fix;

    === FAIL: amd64.integration-cli TestDockerCLICpSuite/TestCpUnprivilegedUser (0.38s)
        docker_cli_cp_test.go:345: assertion failed:
            Command:  su unprivilegeduser -c /usr/local/cli/docker cp cbdd485903262d9de472b0ec6973675f8266209d5f4e32b2b9bd2759fc2d6794:test /tmp/TestDockerCLICpSuiteTestCpUnprivilegedUser231403686/001
            ExitCode: 1
            Error:    exit status 1
            Stdout:
            Stderr:   lstat /tmp/TestDockerCLICpSuiteTestCpUnprivilegedUser231403686/001: permission denied

Signed-off-by: Sebastiaan van Stijn <github@gone.nl>

thaJeztah commented:
Interesting failure on Windows;

```
=== FAIL: github.com/docker/docker/integration/container TestRemoveContainerWithRemovedVolume (0.29s)
    remove_test.go:43: assertion failed: error is not nil: Error response from daemon: invalid volume specification: 'C:\Users\RUNNER~1\AppData\Local\Temp\TestRemoveContainerWithRemovedVolume1128754527\001\hostPath:c:\test': invalid mount config for type "bind": bind source path does not exist: c:\users\runner~1\appdata\local\temp\testremovecontainerwithremovedvolume1128754527\001\hostpath
```

Other failures;

```
=== FAIL: daemon TestTmpfsDevShmNoDupMount (0.01s)
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x28 pc=0x262ed24]

goroutine 147 [running]:
testing.tRunner.func1.2({0x2899000, 0x3e6ae80})
	/usr/local/go/src/testing/testing.go:1526 +0x24e
testing.tRunner.func1()
	/usr/local/go/src/testing/testing.go:1529 +0x39f
panic({0x2899000, 0x3e6ae80})
	/usr/local/go/src/runtime/panic.go:884 +0x213
github.com/docker/docker/daemon.(*fakeImageService).Cleanup(0x0?)
	<autogenerated>:1 +0x24
github.com/docker/docker/daemon.(*Daemon).Shutdown(0xc0009a0500, {0x2e7e4c8?, 0xc000058090?})
	/go/src/github.com/docker/docker/daemon/daemon.go:1292 +0x471
github.com/docker/docker/daemon.setupFakeDaemon.func1()
	/go/src/github.com/docker/docker/daemon/oci_linux_test.go:55 +0x45
testing.(*common).Cleanup.func1()
	/usr/local/go/src/testing/testing.go:1150 +0x11f
testing.(*common).runCleanup(0xc0007489c0, 0x2e54fe0?)
	/usr/local/go/src/testing/testing.go:1328 +0xe2
testing.tRunner.func2()
	/usr/local/go/src/testing/testing.go:1570 +0x29
testing.tRunner(0xc0007489c0, 0x2c50a88)
	/usr/local/go/src/testing/testing.go:1582 +0x135
created by testing.(*T).Run
	/usr/local/go/src/testing/testing.go:1629 +0x3ea

=== FAIL: daemon TestRootMountCleanup (unknown)

=== FAIL: daemon TestVerifyPlatformContainerResources (unknown)

=== FAIL: daemon TestGetBlkioWeightDevices (unknown)

=== FAIL: daemon TestGetBlkioThrottleDevices (unknown)

=== FAIL: pkg/idtools TestNewIDMappings (0.22s)
    idtools_unix_test.go:420: assertion failed: error is not nil: exit status 2: Unable to access /tmp/TestNewIDMappings682566695/001/new-dir directory with user UID:165536 and GID:165536:
        ls: cannot access '/tmp/TestNewIDMappings682566695/001/new-dir': Permission denied
```

```
=== FAIL: amd64.integration-cli TestDockerCLICpSuite/TestCpUnprivilegedUser (0.37s)
    docker_cli_cp_test.go:347: assertion failed: 
        Command:  su unprivilegeduser -c /usr/local/cli-integration/docker cp 95ba7ee40defa94f1534f2ca60e0428ac0cf183f687e5c631964ab1759983b78:test /tmp/TestDockerCLICpSuiteTestCpUnprivilegedUser438127682/001
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   lstat /tmp/TestDockerCLICpSuiteTestCpUnprivilegedUser438127682/001: permission denied
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerCLICpSuite/TestCpUnprivilegedUser (0.37s)

=== FAIL: amd64.integration-cli TestDockerCLICpSuite (21.57s)
```


rootless:

```
=== FAIL: amd64.integration.container TestRemoveContainerWithRemovedVolume (0.10s)
    remove_test.go:43: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume607937866/001/hostPath': mkdir /tmp/TestRemoveContainerWithRemovedVolume607937866/001: permission denied
```

thaJeztah commented:
Failures;

```
=== FAIL: amd64.integration.container TestRemoveContainerWithRemovedVolume (0.04s)
    remove_test.go:40: assertion failed: error is not nil: Error response from daemon: error while creating mount source path '/tmp/TestRemoveContainerWithRemovedVolume3078330942/001/hostPath': mkdir /tmp/TestRemoveContainerWithRemovedVolume3078330942/001: permission denied
```

```
=== FAIL: github.com/docker/docker/integration/container TestRemoveContainerWithRemovedVolume (0.31s)
    remove_test.go:40: assertion failed: error is not nil: Error response from daemon: invalid volume specification: 'C:\Users\RUNNER~1\AppData\Local\Temp\TestRemoveContainerWithRemovedVolume3720226482\001\hostPath:c:\test': invalid mount config for type "bind": bind source path does not exist: c:\users\runner~1\appdata\local\temp\testremovecontainerwithremovedvolume3720226482\001\hostpath

```

```
=== FAIL: amd64.integration-cli TestDockerCLICpSuite/TestCpUnprivilegedUser (0.29s)
    docker_cli_cp_test.go:340: assertion failed: 
        Command:  su unprivilegeduser -c /usr/local/cli-integration/docker cp 80e1a3e0818c1ee53e83e259891cdca90c03b20e7d98a73ec2d9a07d56c2d1ba:test /tmp/TestDockerCLICpSuiteTestCpUnprivilegedUser1338194714/001
        ExitCode: 1
        Error:    exit status 1
        Stdout:   
        Stderr:   lstat /tmp/TestDockerCLICpSuiteTestCpUnprivilegedUser1338194714/001: permission denied
        
        
        Failures:
        ExitCode was 1 expected 0
        Expected no error
    --- FAIL: TestDockerCLICpSuite/TestCpUnprivilegedUser (0.29s)
```
